sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/storage/preference/AndroidPreferenceProvider.kt

```diff
@@ -0,0 +1,140 @@
+ */
+internal class AndroidPreferenceProvider(
+    private val sharedPreferences: SharedPreferences,
+    private val dispatcher: CoroutineDispatcher
+) : PreferenceProvider {
+    override suspend fun hasKey(key: PreferenceKey) =
+        withContext(dispatcher) {
+            sharedPreferences.contains(key.key)
+        }
+
+    @SuppressLint("ApplySharedPref")
+    override suspend fun putString(
+        key: PreferenceKey,
+        value: String?
+    ) = withContext(dispatcher) {
+        val editor = sharedPreferences.edit()
+
+        editor.putString(key.key, value)
+
+        editor.commit()
+
+        Unit
+    }
+
+    override suspend fun getString(key: PreferenceKey) =
+        withContext(dispatcher) {
+            sharedPreferences.getString(key.key, null)
+        }
+
+    override fun observe(key: PreferenceKey): Flow<Unit> =
+        callbackFlow<Unit> {
+            val listener =
+                SharedPreferences.OnSharedPreferenceChangeListener { _, _ ->
+                    // Callback on main thread
+                    trySend(Unit)
+                }
+            sharedPreferences.registerOnSharedPreferenceChangeListener(listener)
+
+            // Kickstart the emissions
+            trySend(Unit)
+
+            awaitClose {
+                sharedPreferences.unregisterOnSharedPreferenceChangeListener(listener)
+            }
+        }.flowOn(dispatcher)
+
+    companion object {
+        suspend fun newStandard(
+            context: Context,
+            filename: String
+        ): PreferenceProvider {
+            /*
+             * Because of this line, we don't want multiple instances of this object created
+             * because we don't clean up the thread afterwards.
+             */
+            val singleThreadedDispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+
+            val sharedPreferences =
+                withContext(singleThreadedDispatcher) {
+                    context.getSharedPreferences(filename, Context.MODE_PRIVATE)
+                }
+
+            return AndroidPreferenceProvider(sharedPreferences, singleThreadedDispatcher)
+        }
+
+        suspend fun newEncrypted(
+            context: Context,
+            filename: String
+        ): PreferenceProvider {
+            /*
+             * Because of this line, we don't want multiple instances of this object created
+             * because we don't clean up the thread afterwards.
+             */
+            val singleThreadedDispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+
+            val mainKey =
+                withContext(singleThreadedDispatcher) {
+                    @Suppress("BlockingMethodInNonBlockingContext")
+                    MasterKey
+                        .Builder(context)
+                        .apply {
+                            setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
+                        }.build()
```

Comment:
I'm not sure I understand exactly what's happening here. Is the key deterministic, or does it vary for each run of the application? How is this key stored?

Reply:
It's a secure and standard way of creating deterministic keys on Android, starting from API 23. The master keys are used to encrypt data encryption keys for encrypting files and preferences. See the EncryptedSharedPreferences [documentation](https://developer.android.com/reference/kotlin/androidx/security/crypto/EncryptedSharedPreferences) for more details. We use the same approach in Demo and Zashi apps.